src/charm.py

```diff
@@ -282,50 +250,50 @@ def _certificate_revoked(self, revoke_list: typing.List[str]) -> None:  # noqa:
             try:
                 self._tls.pop_relation_data_fields(
                     [f"key-{hostname}", f"password-{hostname}"],
-                    tls_certificates_relation,  # type: ignore[arg-type]
+                    tls_requirer_integration,  # type: ignore[arg-type]
                 )
             except KeyError:
                 logger.warning("Relation data for %s already does not exist", hostname)
             self.certificates.request_certificate_revocation(
                 certificate_signing_request=old_csr.encode()
             )
 
+    @validate_config_and_integration(defer=True)
     def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> None:
         """Handle the TLS Certificate invalidation event.
 
         Args:
             event: The event that fires this method.
         """
-        tls_certificates_relation = self._tls.get_tls_relation()
-        if not tls_certificates_relation:
+        try:
+            tls_information = TLSInformation.from_charm(self)
+        except (TlsIntegrationMissingError, InvalidCharmConfigError) as exc:
             self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            logger.error(
```

Comment:
(Please feel free to implement on a followup PR), logger.exception can, by default, print exception stack trace. Hence there won't be a need to pass in exception (`logger.error("<message related to error> %s", exc)` -> `logger.exception("<message related to error>")`)

Reply:
Thanks, this one might be an unintended change when I was keeping all the different PRs in sync, will change this in an upcoming PR.